src/shared/frame-rpc.js

```diff
@@ -21,106 +22,144 @@
+   * `destFrame`.
+   *
+   * @param {Window} sourceFrame
+   * @param {Window} destFrame
+   * @param {string} origin - Origin of destination frame
+   * @param {Record<string, (...args: any[]) => any>} methods - Map of method
+   *   name to method handler
+   */
+  constructor(sourceFrame, destFrame, origin, methods) {
+    this.sourceFrame = sourceFrame;
+    this.destFrame = destFrame;
 
-  this._sequence = 0;
-  this._callbacks = {};
+    if (origin === '*') {
+      this.origin = '*';
+    } else {
+      const uorigin = new URL(origin);
+      this.origin = uorigin.protocol + '//' + uorigin.host;
+    }
 
-  this._onmessage = function (ev) {
-    if (self._destroyed) return;
-    if (self.dst !== ev.source) return;
-    if (self.origin !== '*' && ev.origin !== self.origin) return;
-    if (!ev.data || typeof ev.data !== 'object') return;
-    if (ev.data.protocol !== 'frame-rpc') return;
-    if (!Array.isArray(ev.data.arguments)) return;
-    self._handle(ev.data);
-  };
-  this.src.addEventListener('message', this._onmessage);
-  this._methods =
-    (typeof methods === 'function' ? methods(this) : methods) || {};
-}
+    this._sequence = 0;
+    this._callbacks = {};
 
-RPC.prototype.destroy = function () {
-  this._destroyed = true;
-  this.src.removeEventListener('message', this._onmessage);
-};
+    /** @param {MessageEvent} event */
+    this._onmessage = event => {
+      // Validate message sender and format.
+      if (
+        this._destroyed ||
+        this.destFrame !== event.source ||
+        (this.origin !== '*' && event.origin !== this.origin) ||
+        !event.data ||
+        typeof event.data !== 'object' ||
+        event.data.protocol !== 'frame-rpc' ||
+        !Array.isArray(event.data.arguments)
+      ) {
+        return;
+      }
+      this._handle(event.data);
+    };
+    this.sourceFrame.addEventListener('message', this._onmessage);
+    this._methods = methods;
+  }
 
-/**
- * @param {string} method
- */
-RPC.prototype.call = function (method) {
-  const args = [].slice.call(arguments, 1);
-  return this.apply(method, args);
-};
+  /**
+   * Disconnect the RPC channel. After this is invoked no further method calls
+   * will be received.
+   */
+  destroy() {
+    this._destroyed = true;
+    this.sourceFrame.removeEventListener('message', this._onmessage);
+  }
 
-/**
- * @param {string} method
- * @param {any[]} args
- */
-RPC.prototype.apply = function (method, args) {
-  if (this._destroyed) return;
-  const seq = this._sequence++;
-  if (typeof args[args.length - 1] === 'function') {
-    this._callbacks[seq] = args[args.length - 1];
-    args = args.slice(0, -1);
+  /**
+   * Send an RPC request to the destination frame.
+   *
+   * If the final argument in `args` is a function, it is treated as a callback
+   * which is invoked with the response.
+   *
+   * @param {string} method
+   * @param {any[]} args
+   */
+  call(method, ...args) {
```

Comment:
A curiosity, typescript allows `...rest` parameters at the begging or middle of an assignment:
https://devblogs.microsoft.com/typescript/announcing-typescript-4-2-beta/#leading-middle-rest-elements-in-tuple-types

Reply:
Huh, that's interesting. Good to know!